### ISSUE TITLE
Need a mechanism to deprecate port names

### ISSUE THREAD

**Issue by RussTedrake** [RobotLocomotion/drake#12214]
Currently, I cannot safely change the string name of a port because we do not have a deprecation mechanism.  For instance, MultibodyPlant currently declares an output port "continuous_state", and in #11686 @amcastro-tri points out that we should prefer "state".  But I can't change it safely.

Presumably we need something simple like `DeclareDeprecatedInputPort` (and output).

**Comment by jwnimmer-tri** [comment]
Note that a system can have two output ports that produce the same value, so we could have `"state"` and `"continuous_state"` both offered without recomputing the value.  So we can add the new name, without necessarily removing or deprecating the old name right away.

But as to the request -- when a user requests a port object using its deprecated name, what should happen?  We can't warn at build time.  Do we just want a console log message?

In the past, I have added console warnings; see https://github.com/jwnimmer-tri/drake/commit/4d30ce6a509a3cdc813173db5e346a873ed6f873#diff-e6c2f906908963a5d706dfa0e480652f for example.

For outputs, we can do this pretty easily by having the Calc method for my_deprecated_output log to the console and then Eval the my_nondeprecated_ouput.

For inputs, the System has to decide what happens when none, one, or both of the inputs is connected.  (The warnings may change in the various cases.)  That's best done by some Calc method evaluating all inputs, seeing which are (dis)connected, and dealing from there.

I am not sure how I feel about adding sugar for all of this.  I hope it does not come up very often.

For systems that offer either `get_my_foo_port()` method, or `kFooPortName` constant, I think we should change the string names without a deprecation period (we should still deprecated any functions or constants).  Users should be calling the function or using the constant.  If they call GetOutputPort and manually type in "some_literal_name", then they are doing it wrong and we should worry about breaking their code -- in general I think we should only deprecate things we can control at compile time, and not deprecated magic values checkable only at compile-time.

**Comment by sherm1** [comment]
Trying to handle this with duplicate ports seems like overkill to me and would be especially awkward for input ports. A milder option would be to add a provision for "deprecated names" for existing ports. Using one of those would still dig up the same port but issue a warning message with the preferred name and drop-dead date. Still milder would be to provide "alternate names" for ports and leave the old ones around forever.,

Preferences?

**Comment by jwnimmer-tri** [comment]
I'm sticking with my baseline that we should make no effort to keep the string names intact.

Maybe it would it help if I take the concrete concern from #11686 that inspired this and measure the disruption if we changed it without deprecation?

**Comment by sherm1** [comment]
@RussTedrake how painful would it be just to fix the names cold turkey? We could mention the change on Slack and in release notes. If that's too harsh, what do you think of just permitting alternate names and leaving the old ones in place but undocumented?

**Comment by jwnimmer-tri** [comment]
Having multiple name aliases for the same port means that `system.GetOutputPort('old_foo').get_name()` would return `foo` not `foo_old`, which is still a breaking change, albeit one perhaps less likely to cause disruption.

**Comment by RussTedrake** [comment]
I have quite a different opinion to @jwnimmer-tri ; I think that our adding specific function accessors for every port is not only a lot of boilerplate, but it is done inconsistently throughout the codebase.  I actually wish that the port names were the preferred method for accessing the ports.

My proposal here, though, was that we simply add an e.g. `DeclareDeprecatedInputPort(old_name, new_name, deprecation_date)` method to System; this information would be used *only* when someone calls `GetInputPort(old_name)`... it would print a warning to console and return the reference to the new port.  I wouldn't imaging changing anything deeper in the port code.  I agree that calling `system.GetOutputPort('old_foo').get_name()` would have changed behavior, but at least it would have console output explaining what happened.

**Comment by EricCousineau-TRI** [comment]
Just to add to a "needs-to-be-deprecated" list, `SceneGraph` seems to export `lcm_visualization`, when in fact it's more like `pose_bundle`...

\cc @SeanCurtis-TRI

**Comment by SeanCurtis-TRI** [comment]
I guess that raises the question of "what is the interpretation of a port name"?  Is it syntax or semantics? Its output port is 100% dedicated to lcm visualization. Its *type* happens to be a pose bundle. So, should a port name reflect a type or a semantic interpretation?

 On the other hand, I have the intention of nuking that port when I finally get the chance to rework the visualization.

**Comment by SeanCurtis-TRI** [comment]
However, looking in the review of #12218 I can see where this comment was inspired. We're now connecting the port named `lcm_visualization` to multiple visualizers, only one of which is lcm visualization.

**Comment by SeanCurtis-TRI** [comment]
This conversation never touched the case of *removing* a port. It was primarily about renaming ports.

I'm currently working on deprecating `PoseBundle` (and 95% of what's in `systems/rendering`). Generally, it's straightforward but for three output ports:

  - `SceneGraph::get_pose_bundle_output_port()`
  - `RgbdSensor::X_WB_output_port()` and `RgbdSensorDiscrete::X_WB_output_port()`
  - `ManipulationStation::get_output_port("pose_bundle")

The first two cases are easily resolved by deprecating the methods (if someone is still attempting to access those ports via name or index, then that's on *them*.) The problem is the third case.

It seems the *best* option we have is for me to put a "log once" warning in the calc method for the port. Is that the consensus view?

(Also, I'd be *completely* in favor of getting rid of `MySystem::get_foo_output_port` and replace it with static string `kFooOutputName` and `my_system.get_output_port(MySystem::kFooOutputName)`. The nice thing is those can easily be deprecated and allow for arbitrary re-spellings without hurting anyone.)

**Comment by edrumwri** [comment]
Chiming in as I came across this issue looking at the MBP code for changing "continuous_state" to "state" just to say, I'd be pretty annoyed if get_input_port("portname") suddenly failed one day (after updating my Drake pin) and I had to diagnose the problem and then dig through the code to determine the new string name.

**Comment by RussTedrake** [comment]
@rpoyner-tri -- this seems not very hard to implement, and high value given your naming/renaming exercise.  Would it make sense for you to take this and take it on soon?  (perhaps even before any of the breaking renames land?)

**Comment by rpoyner-tri** [comment]
Let's think about what is needed. Suppose we just go with roughly [Sherm's proposal](https://github.com/RobotLocomotion/drake/issues/12214#issuecomment-543845314), and update it with consequences from recent discussion in #16193.

So, what's on the table?

Get{In,Out}putPort("my_deprecated_name") returns a reference to the renamed port object, issues a deprecation warning (once per process), and that's roughly all it does.  Get("deprecated").port_name() returns the new name. So it goes.

Since ports are inheritable (see for example, VectorSystem), we have to figure out if name deprecations are as well. Likely answer is yes.

Since (so far) @RussTedrake deems naming previously-unnamed Diagram ports to be breaking, we have to be able to handle deprecated names being dependent on system instance data, and getting built/updated at run time.

Since Declare.*Port() are bound for python, we have to likely expose the name deprecation mechanism there as well.

(((Should it bother me that Export.*Port() for Diagram do not seem to be bound for python? Maybe that's its own issue.)))

**Comment by rpoyner-tri** [comment]
Right, and Has.*Port() needs to use the same search code as Get.*Port() (not true presently).

**Comment by rpoyner-tri** [comment]
More fun: does deprecating a default name (previously made with `kUseDefaultName`) on a multi-port system eventually destabilize other default names on that system? Starting with ("u0", "u1") and deprecating "u0" only, do we blunder into having ("newname", "u0") instead of ("newname", "u1")? Do we have to then un-default all input or output ports at once to stabilize names?

**Comment by RussTedrake** [comment]
My 2 cents: I would think we can require that changing one `UseDefaultName` => assigned name requires assigning all subsequent names (in platform review).  Even if the subsequent names are manually assigned to `u4`, `u5`, ...  I would not expect you to handle the cascading default names issue here.

I would also be willing to compromise on the idea that `UseDefaultName` => assigned name constitutes a breaking change, if it's a massive pain point.  I would think that we should (slowly) move towards most ports on `systems/primitives` having all assigned, informative names.

**Comment by rpoyner-tri** [comment]
Part of the difficulty in even talking about this is that kUseDefaultName for diagram ports leads to unspellable hex spam, but kUseDefaultName for leaf ports leads to the u0,u1,... or y0,y1,... sequences.

I'm all for retiring the leaf default names (slowly), and agree that adding some sort of review rule, or other step would probably prevent cascade disasters. I think the diagram default names are almost entirely useless, and should probably not need deprecation/breaking treatment.

**Comment by jwnimmer-tri** [comment]
> kUseDefaultName for diagram ports leads to unspellable hex spam

This is only true when the systems being added to the Diagram do not have any name set.  If the user has given names to their subsystems (which is not at all unusual), then the diagram port names do not have hex spam.

**Comment by rpoyner-tri** [comment]
Right, so that is another run-time case that would have to be handled to "deprecate" kUseDefaultName Diagram port names. I'd love to see an actual in-use program that would break on rename of a kUseDefaultName Diagram port.

**Comment by sherm1** [comment]
And see the venerable issue #7128 for a proposal to fix the awful System naming!

**Comment by rpoyner-tri** [comment]
Well, anyway, I think I've gathered enough fresh opinions to start trying to craft a patch. It may not turn up until January, as time is short and hydro finish work takes precedence. Thanks for the comments all!

**Comment by rpoyner-tri** [comment]
Adding this mechanism has consequences for both deprecation policy (#16159) and tools documentation (#10605).

**Comment by rpoyner-tri** [comment]
Note that the deprecation/breaking logic proposed [here](https://github.com/RobotLocomotion/drake/pull/16193#pullrequestreview-835196314) means that changing the name of a system within a diagram is a breaking change (because it could rename exported ports) and presumably requires deprecation.

**Comment by jwnimmer-tri** [comment]
My solution to this issue is proposed here https://github.com/RobotLocomotion/drake/pull/16268#issuecomment-1006706456:

> (1) for any port, we can mark it "deprecated", which then makes warnings when someone tries to use it.
>
> (2) add sugar for the common case of "I am renaming this port".

The implementation of sugar in (2) looks like:
(a) declare two input ports, one with the old name and one with the new name;
(b) the input port that uses the old name will be marked as deprecated using the mechanism from (1);
(c) provide a cache entry that forwards the value from whichever of the two ports is connected, throwing in case both or neither are connected; during the deprecation period, the user's Calc code will eval using the cache entry instead of the input port.

**Comment by RussTedrake** [comment]
FWIW -- I would love to feel like I could [clean up all of the port names on the manipulator control systems](https://drakedevelopers.slack.com/archives/C2K96TU0K/p1667746668259439), but once again I feel blocked by this missing feature.

And I happened to also stumble on https://github.com/RobotLocomotion/drake/blob/42448c0af1b39f0c46f760e7ae37d77097689ad3/multibody/plant/multibody_plant.cc#L3053-L3054 , because the port name documented in the system doxygen is "continuous_state", but the method to acquire it without using strings it `get_state_output_port()`, not `get_continuous_state_output_port()`.

**Comment by jwnimmer-tri** [comment]
After #18274 lands, I think the only remaining work will be possibly adding sugar for renaming an input port.  It's not clear to me whether the sugar would be a net win, but in any case we can try it and see.

**Comment by EricCousineau-TRI** [comment]
In #18315, I would like to leverage those awesome mechanisms for deprecating ports that are exported within a diagram.
I would assume I can just add a `PassThrough` system, such that I can mock "export" an aliased deprecated port.

Does that sound like a sane method?

**Comment by jwnimmer-tri** [comment]
Closed as completed so far as I know.  If you have a missing use case @EricCousineau-TRI, please open a new ticket with specific details.

**Comment by EricCousineau-TRI** [comment]
Shall do - sorry for delayed response!
